pyzx/editor_actions.py

```diff
@@ -299,6 +300,86 @@ def bialgebra(g: BaseGraph[VT,ET],
+    if len(type1_vertices) <= 1 or len(type2_vertices) <= 1:
+        return None
+    # the vertices must have one external edge
+    for v1 in type1_vertices:
+        if g.vertex_degree(v1) != len(type2_vertices) + 1:
+            return None
+    for v2 in type2_vertices:
+        if g.vertex_degree(v2) != len(type1_vertices) + 1:
+            return None
+    # if all type1 vertices are connected to all type2 vertices with a simple edge, then they are a match
+    for v1 in type1_vertices:
+        for v2 in type2_vertices:
+            edges = list(g.edges(v1, v2))
+            if not (len(edges) == 1 and g.edge_type(edges[0]) == edge_type):
+                return None
+    return type1_vertices, type2_vertices
+
+def bialgebra_op(g: BaseGraph[VT,ET],
+        matches: Tuple[List[VT], List[VT]],
+        edge_type: Optional[EdgeType] = EdgeType.SIMPLE
+        ) -> rules.RewriteOutputType[VT,ET]:
+    def get_neighbors_and_loops(type1_vertices: List[VT], type2_vertices: List[VT]) -> Tuple[List[Tuple[VT, EdgeType]], List[EdgeType]]:
+        neighbors: List[Tuple[VT, EdgeType]] = []
+        loops: List[EdgeType] = []
+        for v1 in type1_vertices:
+            for edge in g.incident_edges(v1):
+                edge_st = g.edge_st(edge)
+                neighbor = edge_st[0] if edge_st[0] != v1 else edge_st[1]
+                if neighbor in type2_vertices:
+                    continue
+                elif neighbor in type1_vertices:
+                    if v1 > neighbor:
+                        loops.append(g.edge_type(edge))
+                else:
+                    neighbors.append((neighbor, g.edge_type(edge)))
+        return neighbors, loops
+
+    def add_vertex_with_averages(vertices, g, vtype):
+        average_row = sum(g.row(v) for v in vertices) / len(vertices)
+        average_qubit = sum(g.qubit(v) for v in vertices) / len(vertices)
+        return g.add_vertex(vtype, average_qubit, average_row)
+
+    def update_etab(etab, new_vertex, neighbors, loops):
+        for n, et in neighbors + [(new_vertex, et) for et in loops]:
+            etab[upair(new_vertex, n)][0 if et == EdgeType.SIMPLE else 1] += 1
+
+    type1_vertices, type2_vertices = matches
+    neighbors1, loops1 = get_neighbors_and_loops(type1_vertices, type2_vertices)
+    neighbors2, loops2 = get_neighbors_and_loops(type2_vertices, type1_vertices)
+
+    new_vertex1 = add_vertex_with_averages(type1_vertices, g, g.type(type2_vertices[0]))
+    new_vertex2 = add_vertex_with_averages(type2_vertices, g, g.type(type1_vertices[0]))
+
+    etab: dict = defaultdict(lambda: [0, 0])
+    if edge_type == EdgeType.SIMPLE:
+        etab[upair(new_vertex1, new_vertex2)] = [1, 0]
+    else:
+        etab[upair(new_vertex1, new_vertex2)] = [0, 1]
+    update_etab(etab, new_vertex1, neighbors1, loops1)
+    update_etab(etab, new_vertex2, neighbors2, loops2)
+
+    g.scalar.add_power(-(len(neighbors1)-1)*(len(neighbors2)-1)) #TODO: not sure if this is correct
```

Comment:
I am not completely sure about the global scalar, so I would appreciate if someone else can confirm

Reply:
This looks about right to me. It should be the inverse of the scalar in the forward bialgebra rule.

Reply:
Okay, do you know what happens with the self-loops? I am not sure if the self-loop scalar is correct even in the forward case.

Reply:
I've never really thought about applying a strong complementarity when there is a self-loop, so I don't know what should happen with the scalar then.

Reply:
I think this is correct so I have removed the todo comment.